pkg/gadgets/socket-collector/tracer/tracer.go

```
@@ -0,0 +1,268 @@
+	}
+
+	return "UNKNOWN"
+}
+
+func getTCPIter() *link.Iter {
+	objs := IterTCPv4Objects{}
+	if err := LoadIterTCPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load TCP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpTcp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach TCP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func getUDPIter() *link.Iter {
+	objs := IterUDPv4Objects{}
+	if err := LoadIterUDPv4Objects(&objs, nil); err != nil {
+		prifixedLog(fmt.Sprintf("Failed to load UDP BPF objects: %v", err))
+		return nil
+	}
+	defer objs.Close()
+
+	it, err := link.AttachIter(link.IterOptions{
+		Program: objs.DumpUdp4,
+	})
+	if err != nil {
+		prifixedLog(fmt.Sprintf("Failed to attach UDP BPF iterator: %v", err))
+		return nil
+	}
+
+	return it
+}
+
+func RunCollector(pid uint32, filter *gadgetv1alpha1.ContainerFilter) (string, error) {
+	if err := increaseRlimit(); err != nil {
+		return "", fmt.Errorf("failed to increase memlock limit: %s", err)
+	}
+
+	// TODO: Use filters to collect only TCP or only UDP information
+	iters := []*link.Iter{}
+	iters = append(iters, getTCPIter())
+	iters = append(iters, getUDPIter())
+
+	// TCP/UDP BPF Iters read information from the current namespace thus
+	// we need to change network namespace before opering the iterator
+	netnsHandle, err := netns.GetFromPid(int(pid))
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve network namespace from PID %d: %w", pid, err)
+	}
+	defer netnsHandle.Close()
+
+	if err := netns.Set(netnsHandle); err != nil {
+		return "", fmt.Errorf("failed to set network namespace of PID %d (NetNsHandle %q): %w", pid, netnsHandle.String(), err)
+	}
+
+	var data []Socket
+
+	for _, it := range iters {
+		if it == nil {
+			continue
+		}
+
+		fd, err := it.Open()
+		if err != nil {
+			prifixedLog(fmt.Sprintf("Failed to open BPF iterator: %v", err))
```

Why is this an error that just gets logged instead of aborting the whole thing like the failure to set the namespace? What could trigger the failure to open this iterator that wouldn't be the case for the rest of the iterators in iters?

---

I did it thinking that if for any reason one of the iterators failed, we will at least show the other ones. Instead, as we discussed, this approach will somehow hide errors to final user. I changed this behaviour and now it will return error.